In [2]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold
import numpy as np
import pickle

def map5eval(preds, actual):
    predicted = preds.argsort(axis=1)[:,-np.arange(6)]
    pr = []
    for pre in predicted:
        pr.append(pre[1:])
    predicted = np.array(pr)
    metric = 0.
    for i in range(5):
        metric += np.sum(actual==predicted[:,i])/(i+1)
    metric /= actual.shape[0]
    return 'MAP@5', metric

path = "vectors1.data"
print(path)
file1 = open(path,"rb")
path = "outputs1.data"
print(path)
file2 = open(path,"rb")
featurevects = pickle.load(file1)
op = pickle.load(file2)
outputs = []
for i in op:
    outputs.append(i[1])

siz = 60000
X = np.array(featurevects[:siz])
y = np.array(outputs[:siz])
zx_test = np.array(featurevects[siz:siz+10000])
zy_test = np.array(outputs[siz:siz+10000])
skf= StratifiedKFold(n_splits=10, shuffle=True, random_state=23)
nest = 300
md = 31
preds = []
for n_fold, (train_idx, valid_idx) in enumerate(skf.split(X,y)):
    X_train, y_train = X[train_idx], y[train_idx]
    X_valid, y_valid = X[valid_idx], y[valid_idx]
    clf = RandomForestClassifier(max_depth=31,n_estimators=300, n_jobs=-1, warm_start=True)
    clf.fit(X_train, y_train)
    preds = clf.predict_proba(X_valid)
    print("fold",n_fold,"- map5:",map5eval(preds,y_valid))

vectors1.data
outputs1.data
fold 0 - map5: 0.49749722222222226
fold 1 - map5: 0.4924083333333333
fold 2 - map5: 0.5036722222222222
fold 3 - map5: 0.49471111111111105
fold 4 - map5: 0.49669722222222223
fold 5 - map5: 0.49912222222222224
fold 6 - map5: 0.48552500000000004
fold 7 - map5: 0.497
fold 8 - map5: 0.4852833333333333
fold 9 - map5: 0.4968


In [ ]:
def map5eval2(preds, actual):
    predicted = []
    for i in range(preds.shape[0]):
        predicted.append(preds[i].argsort()[-5:][::-1])
    predicted = np.array(predicted)
    metric = 0.
    for i in range(5):
        metric += np.sum(actual==predicted[:,i])/(i+1)
        print("recommendation",(i+1),":",np.sum(actual==predicted[:,i]))
    metric /= actual.shape[0]
    return 'MAP@5', -metric

In [ ]:
import pandas as pd
import datetime
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
import numpy as np
import os
import sys
import pickle

def map5eval(preds, actual):
    predicted = preds.argsort(axis=1)[:,-np.arange(6)]
    pr = []
    for pre in predicted:
        pr.append(pre[1:])
    predicted = np.array(pr)
    metric = 0.
    for i in range(5):
        metric += np.sum(actual==predicted[:,i])/(i+1)
    metric /= actual.shape[0]
    return 'MAP@5', metric

def map5eval2(preds, actual):
    predicted = []
    for i in range(preds.shape[0]):
        predicted.append(preds[i].argsort()[-5:][::-1])
    predicted = np.array(predicted)
    metric = 0.
    for i in range(5):
        metric += np.sum(actual==predicted[:,i])/(i+1)
        print("recommendation",(i+1),":",np.sum(actual==predicted[:,i]))
    metric /= actual.shape[0]
    return 'MAP@5', -metric

path = "vectors1.data"
print(path)
file1 = open(path,"rb")
path = "outputs1.data"
print(path)
file2 = open(path,"rb")
featurevects = pickle.load(file1)
op = pickle.load(file2)
outputs = []
for i in op:
    outputs.append(i[1])

n = len(outputs)
print(len(featurevects))
siz = 200000
X = np.array(featurevects[:siz])
y = np.array(outputs[:siz])
zx_test = np.array(featurevects[siz:siz+10000])
zy_test = np.array(outputs[siz:siz+10000])
print("splitting..")
nest = 300
md = 31

rftrain_vect = []
rftrain_op = []

with open("rftrain_vect","rb") as file:
    rftrain_vect = np.array(pickle.load(file))
with open("rftrain_op","rb") as file:
    rftrain_op = np.array(pickle.load(file))

X_test = []
y_test = []

with open("X_test","rb") as file:
    X_test = np.array(pickle.load(file))
with open("y_test","rb") as file:
    y_test = np.array(pickle.load(file))
import sys
test = train_test_split(X,y, test_size=0.2)

clf = RandomForestClassifier(max_depth=md,n_estimators=nest, n_jobs=-1, warm_start=True)
print("fitting..")
clf.fit(rftrain_vect, rftrain_op)
print("predicting..")
# print(X_train.shape,X_valid.shape,zx_test.shape)
preds = clf.predict_proba(X_test)
# preds2 = clf.predict_proba(zx_test)
#     print("max_depth:",md,"--> map5:",map5eval2(preds,y_test)[1])
print("max_depth:",md,"--> map5:",map5eval(preds,y_test)[1])
# print("max_depth:",md,"--> map5:",map5eval(preds2,zy_test)[1])
print("--------------")


# while md<=50:
# break
print("done.")

vectors1.data
outputs1.data
629140
splitting..
fitting..


fold 0 - map5: 0.49749722222222226
***
fold 1 - map5: 0.4924083333333333
***
fold 2 - map5: 0.5036722222222222
***
fold 3 - map5: 0.49471111111111105
***
fold 4 - map5: 0.49669722222222223
***
fold 5 - map5: 0.49912222222222224
***
fold 6 - map5: 0.48552500000000004
***
fold 7 - map5: 0.497
***
fold 8 - map5: 0.4852833333333333
***
fold 9 - map5: 0.4968
***

In [8]:

from sklearn.ensemble import RandomForestClassifier
def map5eval2(preds, actual):
    predicted = []
    for i in range(preds.shape[0]):
        predicted.append(preds[i].argsort()[-5:][::-1])
    predicted = np.array(predicted)
    metric = 0.
    for i in range(5):
        metric += np.sum(actual==predicted[:,i])/(i+1)
    metric /= actual.shape[0]
    return 'MAP@5', -metric
# rftrain_vect = []
# rftrain_op = []
rf_clf = RandomForestClassifier(max_depth=31,n_estimators=300, n_jobs=-1, warm_start=True)
print("fitting..")
rf_clf.fit(rftrain_vect, rftrain_op)
print("predicting..")
rf_preds = rf_clf.predict_proba(X_test)
print("rf_clf --> map5:",map5eval2(preds,y_test)[1])
print("done.")

fitting..
predicting..
rf_clf --> map5: -0.44866888888888884
done.


In [3]:
xgbtrain_vect = []
xgbtrain_op = []
xgbvalid_vect = []
xgbvalid_op = []
X_test = []
y_test = []
rftrain_vect = []
rftrain_op = []
import numpy as np
import pickle
with open("xgbtrain_vect","rb") as file:
    xgbtrain_vect = np.array(pickle.load(file))
with open("xgbtrain_op","rb") as file:
    xgbtrain_op = np.array(pickle.load(file))
with open("xgbvalid_vect","rb") as file:
    xgbvalid_vect = np.array(pickle.load(file))
with open("xgbvalid_op","rb") as file:
    xgbvalid_op = np.array(pickle.load(file))
with open("X_test","rb") as file:
    X_test = np.array(pickle.load(file))
with open("y_test","rb") as file:
    y_test = np.array(pickle.load(file))
with open("rftrain_vect","rb") as file:
    rftrain_vect = np.array(pickle.load(file))
with open("rftrain_op","rb") as file:
    rftrain_op = np.array(pickle.load(file))
